modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/FlowControlledQueueMessageDeliveryImpl.java

```diff
@@ -103,55 +112,58 @@ public int deliverMessageToSubscriptions(StorageQueue storageQueue) throws Andes
                         }
 
                         message.markAsScheduledToDeliver(localSubscription);
+                        //if the message was delivered, it needs to be removed
                         iterator.remove();
                         MessageFlusher.getInstance().deliverMessageAsynchronously(localSubscription, message);
                         numOfCurrentMsgDeliverySchedules++;
 
                         //for queue messages and durable topic messages (as they are now queue messages)
                         // we only send to one selected subscriber if it is a queue message
-                        break;
-                    }
-                }
-
-                //if the message was delivered, it needs to be removed
-                if (numOfCurrentMsgDeliverySchedules == 1) {
-                    if (log.isDebugEnabled()) {
-                        log.debug("Removing Scheduled to send message from buffer. MsgId= " + message.getMessageID());
-                    }
-                    sentMessageCount++;
-                } else {
-
-                    //if no subscriber has a matching selector, route message to DLC queue
-                    if (!subscriberWithMatchingSelectorFound && !suspendedSubFound) {
```

Comment:
If a message does not match any selector, we don't delete it now, but keep it stored since a consumer could later come without any selectors, is it?